stdlib/dynarray.mli

```diff
@@ -0,0 +1,634 @@
+    providing fast access to elements at arbitrary positions.
+
+{[
+    let dynarray_of_list li =
+      let arr = Dynarray.create () in
+      List.iter (fun v -> Dynarray.add_last arr v) li;
+      arr
+]}
+
+    The {!Buffer} module provides similar features, but it is
+    specialized for accumulating characters into a dynamically-resized
+    string.
+
+    The {!Stack} module provides a last-in first-out data structure
+    that can be easily implemented on top of dynamic arrays.
+
+    {b Warning.} In their current implementation, the memory layout
+    of dynamic arrays differs from the one of {!Array}s. See the
+    {{!section:memory_layout} Memory Layout} section for more information.
+
+    @since 5.2
+*)
+
+(** {b Unsynchronized accesses} *)
+
+[@@@alert unsynchronized_access
+    "Unsynchronized accesses to dynamic arrays are a programming error."
+]
+
+(**
+   Unsynchronized accesses to a dynamic array may lead to an invalid
+   dynamic array state. Thus, concurrent accesses to dynamic arrays
+   must be synchronized (for instance with a {!Mutex.t}).
+*)
+
+(** {1:dynarrays Dynamic arrays} *)
+
+type 'a t
+(** A dynamic array containing values of type ['a].
+
+    A dynamic array [a] provides constant-time [get] and [set]
+    operations on indices between [0] and [Dynarray.length a - 1]
+    included. Its {!length} may change over time by adding or removing
+    elements to the end of the array.
+
+    We say that an index into a dynarray [a] is valid if it is in
+    [0 .. length a - 1] and invalid otherwise.
+*)
+
+val create : unit -> 'a t
+(** [create ()] is a new, empty array. *)
+
+val make : int -> 'a -> 'a t
+(** [make n x] is a new array of length [n], filled with [x].
+
+    @raise Invalid_argument if [n < 0] or [n > Sys.max_array_length].
+*)
+
+val init : int -> (int -> 'a) -> 'a t
+(** [init n f] is a new array [a] of length [n],
+    such that [get a i] is [f i]. In other words,
+    the elements of [a] are [f 0], then [f 1],
+    then [f 2]... and [f (n - 1)] last, evaluated
+    in that order.
+
+    This is similar to {!Array.init}.
+
+    @raise Invalid_argument if [n < 0] or [n > Sys.max_array_length].
+*)
+
+val get : 'a t -> int -> 'a
+(** [get a i] is the [i]-th element of [a], starting with index [0].
+
+    @raise Invalid_argument if the index is invalid *)
+
+val set : 'a t -> int -> 'a -> unit
+(** [set a i x] sets the [i]-th element of [a] to be [x].
+
+    [i] must be a valid index. [set] does not add new elements to the
+    array -- see {!add_last} to add an element.
+
+    @raise Invalid_argument if the index is invalid. *)
+
+val length : 'a t -> int
+(** [length a] is the number of elements in the array. *)
+
+val is_empty : 'a t -> bool
+(** [is_empty a] is [true] if [a] is empty, that is, if [length a = 0]. *)
+
+val get_last : 'a t -> 'a
+(** [get_last a] is the element of [a] at index [length a - 1].
+
+    @raise Invalid_argument if [a] is empty.
```

Comment:
@dbuenzli suggested Invalid_argument on empty dynarrays, but I hadn't thought of the inconsistency with pop_last. I changed to use `Not_found` as well. Thanks!

Reply:
And now you are inconsistent with `get`… If you ask me I'd rather have the equation:

```
get_last a = get a (length a - 1)
```

Reply:
Also you are setting an unwanted precedent. 

To the best of my knowledge no `get_*` function of the `Stdlib` ever raises `Not_found` (thank god!). 

I'd be happy if we can keep the idea that `get_*` functions are for when we know the element is there and thus a programming error (and thus `Invalid_argument`), for when it's not.

Reply:
So should we consider renaming `get_last_opt` into `find_last`?

Reply:
Yes, either rename `get_last` or make it not raise `Not_found`. (And either rename `get_last_opt` or remove it for consistency with the absence of `Not_found`, since there is no corresponding `get_opt`... Maybe that's what confused me initially.) At this point the choice is up to you.

Reply:
The new iteration has

```ocaml
val get_last : 'a t -> 'a
(** [get_last a] is the element of [a] at index [length a - 1].

    @raise Invalid_argument if [a] is empty.
*)

val find_last : 'a t -> 'a option
(** [find_last a] is [None] if [a] is empty
    and [Some (get_last a)] otherwise. *)
```